L1Trigger/TrackerTFP/src/LayerEncoding.cc

```diff
@@ -0,0 +1,94 @@
+    auto equalRZ = [](const SensorModule* lhs, const SensorModule* rhs){ return abs(lhs->r() - rhs->r()) < 1.e-3 && abs(lhs->z() - rhs->z()) < 1.e-3; };
+    stable_sort(sensorModules.begin(), sensorModules.end(), smallerR);
+    stable_sort(sensorModules.begin(), sensorModules.end(), smallerZ);
+    sensorModules.erase(unique(sensorModules.begin(), sensorModules.end(), equalRZ), sensorModules.end());
+    for (int binEta = 0; binEta < setup_->numSectorsEta(); binEta++) {
+      const double sectorCot = (sinh(setup_->boundarieEta(binEta + 1)) + sinh(setup_->boundarieEta(binEta))) / 2.;
+      const double sectorZT = sectorCot * setup_->chosenRofZ();
+      const double rangeZT = (sinh(setup_->boundarieEta(binEta + 1)) - sinh(setup_->boundarieEta(binEta))) / 2. * setup_->chosenRofZ();
+      for (int binZT = 0; binZT < pow(2, zT_->width()); binZT++) {
+        const double zT = zT_->floating(zT_->toSigned(binZT));
+        if (abs(zT) > rangeZT + zT_->base() / 2.)
+          continue;
+        for (int binCot = 0; binCot < pow(2, cot_->width()); binCot++) {
+          const double cot = cot_->floating(cot_->toSigned(binCot));
+          const double z0 = zT - cot * setup_->chosenRofZ();
+          if (abs(z0) > setup_->beamWindowZ() + cot_->base() * setup_->chosenRofZ() / 2.)
+            continue;
+          vector<set<int>> layers(2);
+          const vector<double> zTs = {sectorZT + zT - zT_->base() / 2., sectorZT + zT + zT_->base() / 2.};
+          const vector<double> cots = {sectorCot + cot - cot_->base() / 2., sectorCot + cot + cot_->base() / 2.};
+          for (const SensorModule* sm : sensorModules) {
+            for (int i = 0; i < boundaries; i++) {
+              const int j = boundaries - i - 1;
+              const double zTi = zTs[sm->r() > setup_->chosenRofZ() ? i : j];
+              const double coti = cots[sm->r() > setup_->chosenRofZ() ? j : i];
+              const double d = (zTi - sm->z() + (sm->r() - setup_->chosenRofZ()) * coti) / (sm->cos() - sm->sin() * coti);
+              if (abs(d) < sm->numColumns() * sm->pitchCol() / 2.)
+                layers[i].insert(sm->layerId());
+            }
+          }
+          set<int> maybeLayer;
+          set_symmetric_difference(layers[0].begin(), layers[0].end(), layers[1].begin(), layers[1].end(), inserter(maybeLayer, maybeLayer.end()));
+          set<int> layerEncoding;
+          set_union(layers[0].begin(), layers[0].end(), layers[1].begin(), layers[1].end(), inserter(layerEncoding, layerEncoding.end()));
+          layerEncoding_[binEta][binZT][binCot] = vector<int>(layerEncoding.begin(), layerEncoding.end());
+          maybeLayer_[binEta][binZT][binCot] = vector<int>(maybeLayer.begin(), maybeLayer.end());
+        }
+      }
+    }
+  }
+  
+  const int LayerEncoding::layerIdKF(int binEta, int binZT, int binCot, int layerId) const {
+    const vector<int>& layers = layerEncoding_[binEta][binZT][binCot];
```

Comment:
Add comment explaining how KF layer ID is defined. e.g. "Set of layers in each (zT,tanL) digi Bin of each eta sector numbered 0->N".

Reply:
done

Reply:
Still seems vague. Doesn't say how the layer is encoded.